core/src/main/java/io/grpc/internal/TransportTracer.java

```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright 2017, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.internal;
+
+import com.google.common.base.Preconditions;
+import io.grpc.Status;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * A class for gathering statistics about a transport. This is an experimental feature.
+ * Can only be called from the transport thread unless otherwise noted.
+ */
+public final class TransportTracer {
+  private long streamsStarted;
+  private long lastStreamCreatedTimeNanos;
+  private long streamsSucceeded;
+  private long streamsFailed;
+  private long keepAlivesSent;
+  private FlowControlReader flowControlWindowReader;
+
+  // TODO(zpencer): msg sent can piggyback on framer's writequeue and avoid syncing the update
+  private final LongCounter messagesSent = LongCounterFactory.create();
+  private volatile long lastMessageSentTimeNanos;
```

Comment:
non volatile

Reply:
This is currently happening in the framer, which is in the application thread(s). I will follow up with a PR to move this counter increment to the transport thread, since this PR is already big enough.